radicle-tracker/src/metadata/clock.rs

```diff
@@ -0,0 +1,263 @@
+//! #
+//! #     Ok(())
+//! # }
+//! ```
+
+// Rough calculations for the number of seconds in some larger unit
+const SECONDS_IN_MINUTE: u64 = 60;
+const SECONDS_IN_HOUR: u64 = SECONDS_IN_MINUTE * 60;
+const SECONDS_IN_DAY: u64 = SECONDS_IN_HOUR * 24;
+const SECONDS_IN_WEEK: u64 = SECONDS_IN_DAY * 7;
+const SECONDS_IN_MONTH: u64 = SECONDS_IN_WEEK * 4;
+const SECONDS_IN_YEAR: u64 = SECONDS_IN_MONTH * 12;
+
+use std::time::{Duration, SystemTime, SystemTimeError};
+
+/// The elapsed time from some previous moment in the past. This is to capture
+/// concepts like, "this comment was posted 4 minutes ago", or "5 days ago", or
+/// "1 year ago".
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Elapsed {
+    /// Elapsed "x minutes ago".
+    Minutes(u64),
+    /// Elapsed "x hours ago".
+    Hours(u64),
+    /// Elapsed "x days ago".
+    Days(u64),
+    /// Elapsed "x weeks ago".
+    Weeks(u64),
+    /// Elapsed "x months ago".
+    Months(u64),
+    /// Elapsed "x years ago".
+    Years(u64),
+}
+
+impl std::fmt::Display for Elapsed {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Elapsed::Minutes(m) => write!(f, "{} minutes ago", m),
+            Elapsed::Hours(m) => write!(f, "{} hours ago", m),
+            Elapsed::Days(m) => write!(f, "{} days ago", m),
+            Elapsed::Weeks(m) => write!(f, "{} weeks ago", m),
+            Elapsed::Months(m) => write!(f, "{} months ago", m),
+            Elapsed::Years(m) => write!(f, "{} years ago", m),
+        }
+    }
+}
+
+/// `Clock` captures the minimal functionality for grabbing a point in time and
+/// calculating the duration since another point in time.
+pub trait Clock {
+    /// Time calculations tend to return errors.
+    type Error;
+
+    /// Get the current timestamp.
+    fn current_time() -> Self;
+
+    /// Get the duration for a timestamp in the past, `self`, up to the point of
+    /// `other`. It should return [`std::time::Duration`], which can be used
+    /// for further calculation.
+    fn duration_since(&self, other: &Self) -> Result<Duration, Self::Error>;
+}
+
+/// A minimal clock for getting a moment in time and calculating the duration up
+/// to another moment in time.
+///
+/// It does this through its instance of [`Clock`] trait.
+///
+/// For usability, it also provides a function [`RadicleClock::elapsed`] which
+/// returns an [`Elapsed`] value for presentation purposes.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct RadicleClock(SystemTime);
+
+impl Clock for RadicleClock {
+    type Error = SystemTimeError;
+
+    fn current_time() -> Self {
+        RadicleClock(SystemTime::now())
+    }
+
+    fn duration_since(&self, other: &Self) -> Result<Duration, Self::Error> {
+        let timestamp = self.0.duration_since(SystemTime::UNIX_EPOCH)?;
+        let other_timestamp = other.0.duration_since(SystemTime::UNIX_EPOCH)?;
+
+        Ok(other_timestamp - timestamp)
```

Comment:
Something I'm not sure about is that this will panic if they're not ordered right :/

Reply:
Should we take the `abs` value or introduce an `enum`, i.e.
```rust
pub enum When<A> {
  Before(A),
  After(A),
}
```

Reply:
Yeah, the Rust time API is pretty strange. I'm guessing that it is like this because `std` does not have arbitrary precision arithmetic types -- for whatever reason. Luckily, there's a crate for it!
I think we can make all operations total, and more like Haskell `time`, which gets it right:

```rust
use std::{
    fmt,
    ops::{Neg, Sub},
    thread,
    time::{Duration, SystemTime, UNIX_EPOCH},
};

use num_bigint::BigInt;

#[derive(Debug)]
struct TimeDiff(BigInt);

impl fmt::Display for TimeDiff {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl Sub for TimeDiff {
    type Output = TimeDiff;

    fn sub(self, rhs: TimeDiff) -> Self::Output {
        TimeDiff(self.0 - rhs.0)
    }
}

impl From<SystemTime> for TimeDiff {
    fn from(t: SystemTime) -> Self {
        Self(
            t.duration_since(UNIX_EPOCH)
                .map(|d| d.as_secs().into())
                .unwrap_or_else(|e| BigInt::from(e.duration().as_secs()).neg()),
        )
    }
}

trait Clock {
    fn current_time() -> Self;
    fn diff_since(&self, other: &Self) -> TimeDiff;
}

#[derive(Debug)]
struct RadClock(SystemTime);

impl Clock for RadClock {
    fn current_time() -> Self {
        Self(SystemTime::now())
    }

    fn diff_since(&self, other: &Self) -> TimeDiff {
        TimeDiff::from(self.0) - TimeDiff::from(other.0)
    }
}

fn main() {
    let now = RadClock::current_time();
    println!("now: {:?}", now);

    thread::sleep(Duration::from_secs(1));

    let later = RadClock::current_time();
    println!("later: {:?}", later);

    println!("diff: {}", later.diff_since(&now));
}
```

Reply:
Nice :o told you you're a smart :cookie:

One thing I'm not sure about is how the end-user of this is supposed to determine which time came later though? Or does it even matter :thinking:

Thinking about it, I suppose not. Basically they'll grab a comment and do `RadClock::current_time().diff_since(comment.timestamp())` to display the difference. #rubberduck :duck:

Reply:
Ok, I ran with it. This looks better now, imho. Lemme know if you see anything funny